### PR TITLE
Feat(Inventory): Basic auth for Agent

### PR DESCRIPTION
**Pull request by MyvTsv**
## Checklist before requesting a review

*Please delete options that are not relevant.*

- [x] I have read the CONTRIBUTING document.
- [x] I have performed a self-review of my code.
- [x] I have added tests that prove my fix is effective or that my feature works.

## Description

Implement Basic Auth form for inventory endpoint

## Screenshots (if appropriate):

Capture d’écran du 2024-09-11 15-00-48

### CODE REVIEW

File: src/Glpi/Inventory/Conf.php [hunk 1]
```diff
@@ -384,26 +384,63 @@ public function showConfigForm()
...
             echo "<tr class='tab_bg_1'>";
             echo "<td>";
             echo "<label for='auth'>" . __s('Authorization header') . "</label>";
             echo "</td>";
             echo "<td>";
             Dropdown::showFromArray('auth_required', [
                 'none' => __('None'),
-                'client_credentials' => __('OAuth - Client credentials')
+                'client_credentials' => __('OAuth - Client credentials'),
+                'basic_authentification' => __('Basic Authentification'),
```
Comment (stonebuzz) [inline src/Glpi/Inventory/Conf.php:395]: Can you move `client_credentials` and `basic_authentification` as `const`

```php
const CLIENT_CREDENTIALS = 'client_credentials';
const BASIC_AUTH = 'basic_auth';
```

allows us (among other things) to change only one place (the day we want to update the value of the constant).

File: src/Glpi/Inventory/Conf.php [hunk 2]
```diff
@@ -384,26 +384,63 @@ public function showConfigForm()
...
+            echo "</td>";
+            echo "<td>";
+            echo Html::input("basic_auth_login", [
+                "value" => $config["basic_auth_login"]
+            ]);
+            echo "</td>";
+            echo "</tr>";
+            echo "<tr class='tab_bg_1' id='basic_auth_password_row'>";
+            echo "<td>";
+            echo "<label for='basic_auth_password'>" . __s('Password') . "</label>";
```
Comment (stonebuzz) [inline src/Glpi/Inventory/Conf.php:412]:
```suggestion
            echo "<label for='basic_auth_password'>" . __('Password') . "</label>";
```
`__s('string')` escape HTML entities (no need here)

File: src/Glpi/Inventory/Conf.php [hunk 3]
```diff
@@ -384,26 +384,63 @@ public function showConfigForm()
...
-                'client_credentials' => __('OAuth - Client credentials')
+                'client_credentials' => __('OAuth - Client credentials'),
+                'basic_authentification' => __('Basic Authentification'),
             ], [
                 'value' => $config['auth_required'] ?? 'none'
             ]);
             echo "</td></tr>";
+            echo "<tr class='tab_bg_1' id='basic_auth_login_row'>";
+            echo "<td>";
+            echo "<label for='basic_auth_login'>" . __s('Login') . "</label>";
```
Comment (stonebuzz) [inline src/Glpi/Inventory/Conf.php:402]:
```suggestion
            echo "<label for='basic_auth_login'>" . __('Login') . "</label>";
```

`__s('string')` escape HTML entities (no need here)

Comment (cedric-anne) [reply src/Glpi/Inventory/Conf.php:402]: It cannot hurts.

File: src/Glpi/Inventory/Conf.php (hunk 1 above)
Comment (cconard96) [inline src/Glpi/Inventory/Conf.php:395]: All uses of "Authentification" should be "Authentication" instead.

File: src/Glpi/Agent/Communication/AbstractRequest.php [hunk 4]
```diff
@@ -223,6 +223,10 @@ public function handleRequest($data): bool
             }
         }
 
+        /*if ($auth_required === 'basic_authentification') {
+
+        }*/
+
```
Comment (trasher) [inline src/Glpi/Agent/Communication/AbstractRequest.php:229]:
```suggestion
```

File: front/inventory.conf.php [hunk 5]
```diff
@@ -40,11 +40,26 @@
...
+        ) && $_POST['auth_required'] === Conf::BASIC_AUTH
+    ) {
+        Session::addMessageAfterRedirect(
+            "Basic Authentication is active. The login and/or password fields are missing.",
+            false,
+            ERROR
+        );
+        Html::back();
+    }
+    $_POST['basic_auth_password'] = $glpikey->encrypt($_POST['basic_auth_password']);
```
Comment (stonebuzz) [inline front/inventory.conf.php:62]: This part should be place in  `src/Inventory/Conf.php` (`saveConf()`)

And return `false` if `login` or `pwd` are empty

File: front/inventory.conf.php [hunk 6]
```diff
@@ -40,11 +40,26 @@
...
     $conf->saveConf($_POST);
     Session::addMessageAfterRedirect(
```
Comment (stonebuzz) [inline front/inventory.conf.php:64]: `Session::addMessageAfterRedirect` should be called only if 

`$conf->saveConf($_POST)` return `true`

File: src/Glpi/Agent/Communication/AbstractRequest.php [hunk 7]
```diff
@@ -223,6 +225,29 @@ public function handleRequest($data): bool
             }
         }
 
+        if ($auth_required === Conf::BASIC_AUTH) {
+            $authorization = $this->headers->getHeader('Authorization');
+            if (!$authorization) {
```
Comment (stonebuzz) [inline src/Glpi/Agent/Communication/AbstractRequest.php:230]:
```suggestion
            if (is_null($authorization)) {
```

Because `getHeader()` return `null` if related header name not exist

File: src/Glpi/Agent/Communication/AbstractRequest.php [hunk 8]
```diff
@@ -223,6 +225,29 @@ public function handleRequest($data): bool
...
+                $passwordAgent = \Config::getConfigurationValue('inventory', 'basic_auth_password');
+                $authData = base64_decode(substr($authorization, 6));
+                list($loginUser, $passwordUser) = explode(':', $authData, 2);
+                if (
+                    $loginUser !== $loginAgent ||
+                    $passwordUser !== $glpikey->decrypt($passwordAgent)
+                ) {
+                    $this->setMode(self::JSON_MODE);
+                    $this->addError('Acces denied. Login or password is wrong.', 401);
+                }
```
Comment (stonebuzz) [inline src/Glpi/Agent/Communication/AbstractRequest.php:247]:
```suggestion
                $allowed = false;
                // if Authorization start with 'Basic'
                if (preg_match('/^Basic\s+(.*)$/i', $authorizationHeader, $matches)) {
                    $inventory_login = \Config::getConfigurationValue('inventory', 'basic_auth_login');
                    $inventory_password = (new GLPIKey())->decrypt(\Config::getConfigurationValue('inventory', 'basic_auth_password'));
                    $agent_credential = $base64_decode($matches[1]);
                    list($agent_login, $agent_password) = explode(':', $agent_credential, 2);
                    if (
                        $inventory_login == $agent_login &&
                        $inventory_password == agent_password
                    ) {
                        $allowed = true;
                    }
                }

                if (!$allowed) {
                    $this->setMode(self::JSON_MODE);
                    $this->addError('Access denied. Wrong login or password for basic authentication.', 401);
                    return false;
                }
```

You should use `snake_case` for variable naming

(This naming convention should be applied to the rest of the PR)

File: src/Glpi/Inventory/Conf.php [hunk 9]
```diff
@@ -384,26 +391,69 @@ public function showConfigForm()
...
+            echo "<tr class='tab_bg_1' id='basic_auth_password_row'>";
+            echo "<td>";
+            echo "<label for='basic_auth_password'>"
+                . __('Password') .
+                "<span class='required'>*</span>
+                </label>";
+            echo "</td>";
+            echo "<td>";
+            echo Html::input("basic_auth_password", [
+                "value" => $config["basic_auth_password"],
```
Comment (stonebuzz) [inline src/Glpi/Inventory/Conf.php:429]:
```suggestion
                "value" => (new GLPIKey())->decrypt($config["basic_auth_password"]),
```

File: src/Glpi/Inventory/Conf.php [hunk 10]
```diff
@@ -344,7 +349,9 @@ public function showConfigForm()
          */
         global $CFG_GLPI, $PLUGIN_HOOKS;
 
+        $glpikey = new GLPIKey();
```
Comment (stonebuzz) [inline src/Glpi/Inventory/Conf.php:352]:
```suggestion
```

File: src/Glpi/Inventory/Conf.php [hunk 11]
```diff
@@ -344,7 +349,9 @@ public function showConfigForm()
...
         $config = \Config::getConfigurationValues('inventory');
+        $config['basic_auth_password'] = $glpikey->decrypt($config['basic_auth_password']);
```
Comment (stonebuzz) [inline src/Glpi/Inventory/Conf.php:354]:
```suggestion
```

File: front/inventory.conf.php [hunk 12]
```diff
@@ -40,11 +40,26 @@
 Html::header(__('Inventory'), $_SERVER['PHP_SELF'], "admin", "glpi\inventory\inventory");
 
 $conf = new Conf();
+$glpikey = new GLPIKey();
```
Comment (stonebuzz) [inline front/inventory.conf.php:43]:
```suggestion
```

See my next suggestion

File: src/Glpi/Inventory/Conf.php [hunk 13]
```diff
@@ -384,26 +391,69 @@ public function showConfigForm()
...
+                "value" => $config["basic_auth_login"],
+            ]);
+            echo "</td>";
+            echo "</tr>";
+            echo "<tr class='tab_bg_1' id='basic_auth_password_row'>";
+            echo "<td>";
+            echo "<label for='basic_auth_password'>"
+                . __('Password') .
+                "<span class='required'>*</span>
+                </label>";
```
Comment (stonebuzz) [inline src/Glpi/Inventory/Conf.php:425]:
```suggestion
        echo __('Password');
        echo "<span class='required'>*</span>";
        echo "</label>";
```

File: src/Glpi/Inventory/Conf.php [hunk 14]
```diff
@@ -384,26 +391,69 @@ public function showConfigForm()
...
             ], [
                 'value' => $config['auth_required'] ?? 'none'
             ]);
             echo "</td></tr>";
+            echo "<tr class='tab_bg_1' id='basic_auth_login_row'>";
+            echo "<td>";
+            echo "<label for='basic_auth_login'>"
+                . __('Login') .
+                "<span class='required'>*</span>
+                </label>";
```
Comment (stonebuzz) [inline src/Glpi/Inventory/Conf.php:412]:
```suggestion
        echo __('Login');
        echo "<span class='required'>*</span>";
        echo "</label>";
```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: front/inventory.conf.php [hunk 15]
```diff
@@ -40,11 +40,26 @@
...
 } elseif (isset($_POST['update'])) {
     unset($_POST['update']);
+    if (
+        (
+            !$_POST['basic_auth_password'] ||
+            !$_POST['basic_auth_login']
+        ) && $_POST['auth_required'] === Conf::BASIC_AUTH
+    ) {
+        Session::addMessageAfterRedirect(
+            "Basic Authentication is active. The login and/or password fields are missing.",
```
Comment (cconard96) [inline front/inventory.conf.php:56]: Must be translatable and use function "__s".

File: src/Glpi/Inventory/Conf.php [hunk 16]
```diff
@@ -384,26 +391,69 @@ public function showConfigForm()
...
+                self::CLIENT_CREDENTIALS => __('OAuth - Client credentials'),
+                self::BASIC_AUTH => __('Basic Authentication'),
             ], [
                 'value' => $config['auth_required'] ?? 'none'
             ]);
             echo "</td></tr>";
+            echo "<tr class='tab_bg_1' id='basic_auth_login_row'>";
+            echo "<td>";
+            echo "<label for='basic_auth_login'>"
+                . __('Login') .
```
[the remark this answers is not in the document]
Comment (stonebuzz) [reply src/Glpi/Inventory/Conf.php:410]: even if they are no HTML entities?

[the remark this answers is not in the document]
Comment (cconard96) [reply src/Glpi/Inventory/Conf.php:410]: It is the safest option. Who knows what different languages will have in their translations and it isnt like they are strictly reviewed.

[the remark this answers is not in the document]
Comment (stonebuzz) [reply src/Glpi/Inventory/Conf.php:410]: :+1:

File: front/inventory.conf.php [hunk 17]
```diff
@@ -45,12 +45,14 @@
...
-        INFO
-    );
+    $conf_is_success = $conf->saveConf($_POST);
+    if ($conf_is_success) {
+        Session::addMessageAfterRedirect(
+            __s('Configuration has been updated'),
+            false,
+            INFO
+        );
+    }
```
Comment (stonebuzz) [inline front/inventory.conf.php:55]:
```suggestion
    if ($conf->saveConf($_POST)) {
        Session::addMessageAfterRedirect(
            __s('Configuration has been updated'),
            false,
            INFO
        );
    }
```

File: src/Glpi/Agent/Communication/Headers/Common.php [hunk 18]
```diff
@@ -40,6 +40,7 @@
 
 class Common
 {
+
```
Comment (stonebuzz) [inline src/Glpi/Agent/Communication/Headers/Common.php:43]:
```suggestion
```

File: src/Glpi/Inventory/Conf.php [hunk 19]
```diff
@@ -1102,6 +1150,24 @@ public function saveConf(array $values)
...
+                __s("Basic Authentication is active. The login and/or password fields are missing."),
+                false,
+                ERROR
+            );
+            return false;
+        }
+
+        if (!is_null($values['basic_auth_password'])) {
+            $values['basic_auth_password'] = (new GLPIKey())->encrypt($values['basic_auth_password']);
+        }
```
Comment (stonebuzz) [inline src/Glpi/Inventory/Conf.php:1169]:
```suggestion
        if (
            $values['auth_required'] === Conf::BASIC_AUTH
            && !empty($values['basic_auth_password'])
            && !empty($values['basic_auth_login'])
        ) {
            $values['basic_auth_password'] = (new GLPIKey())->encrypt($values['basic_auth_password']);
        } else {
            Session::addMessageAfterRedirect(
                __s("Login and password fields cannot be empty for basic authentication."),
                false,
                ERROR
            );
            return false;
        }

```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: src/Glpi/Inventory/Conf.php [hunk 20]
```diff
@@ -1102,6 +1150,21 @@ public function saveConf(array $values)
...
+        ) {
+            $values['basic_auth_password'] = (new GLPIKey())->encrypt($values['basic_auth_password']);
+        } else {
+            Session::addMessageAfterRedirect(
+                __s("Basic Authentication is active. The login and/or password fields are missing."),
+                false,
+                ERROR
+            );
+            return false;
+        }
```
Comment (stonebuzz) [inline src/Glpi/Inventory/Conf.php:1166]:
```suggestion
        if ($values['auth_required'] === Conf::BASIC_AUTH) {
            if (
                    !empty($values['basic_auth_password']) &&
                    !empty($values['basic_auth_login'])
            ) {
                $values['basic_auth_password'] = (new GLPIKey())->encrypt($values['basic_auth_password']);
            } else {
                Session::addMessageAfterRedirect(
                    __s("Basic Authentication is active. The login and/or password fields are missing."),
                    false,
                    ERROR
                );
                return false;
            }
        }
```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: src/Glpi/Inventory/Conf.php [hunk 21]
```diff
@@ -1102,6 +1150,22 @@ public function saveConf(array $values)
             $values['stale_agents_status_condition'] = ['all'];
         }
 
+        if ($values['auth_required'] === Conf::BASIC_AUTH) {
```
Comment (stonebuzz) [inline src/Glpi/Inventory/Conf.php:1153]:
```suggestion
        if (isset($values['auth_required']) && $values['auth_required'] === Conf::BASIC_AUTH) {
```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: tests/web/Glpi/Inventory/Request.php [hunk 22]
```diff
@@ -207,4 +207,195 @@ public function testPrologRequest()
...
+                '<REQUEST>' .
+                  '<DEVICEID>mydeviceuniqueid</DEVICEID>' .
+                  '<QUERY>PROLOG</QUERY>' .
+                '</REQUEST>'
+            ]
+        );
+        $this->checkXmlResponse($res, '<PROLOG_FREQ>24</PROLOG_FREQ><RESPONSE>SEND</RESPONSE>', 200);
+    }
+
+    public function testAuthWithoutBasic()
```
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:270]:
```suggestion
    public function testAuthBasicMalformed()
```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: src/Glpi/Inventory/Conf.php [hunk 23]
```diff
@@ -384,26 +389,69 @@ public function showConfigForm()
...
             echo "<tr class='tab_bg_1'>";
             echo "<td>";
             echo "<label for='auth'>" . __s('Authorization header') . "</label>";
             echo "</td>";
             echo "<td>";
             Dropdown::showFromArray('auth_required', [
                 'none' => __('None'),
-                'client_credentials' => __('OAuth - Client credentials')
+                self::CLIENT_CREDENTIALS => __('OAuth - Client credentials'),
+                self::BASIC_AUTH => __('Basic Authentication'),
```
Comment (stonebuzz) [inline src/Glpi/Inventory/Conf.php:400]:
```suggestion
                self::CLIENT_CREDENTIALS => __s('OAuth - Client credentials'),
                self::BASIC_AUTH => __s('Basic Authentication'),
```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: src/Glpi/Inventory/Conf.php [hunk 24]
```diff
@@ -384,26 +389,69 @@ public function showConfigForm()
             echo '</div>';
             echo "</td>";
             echo "</tr>";
-
             echo "<tr class='tab_bg_1'>";
             echo "<td>";
-            echo "<label for='auth'>" . __s('Authorization header') . "</label>";
+            echo CommonGLPI::createTabEntry(__s('Authorization header'), 0, null, 'ti ti-cloud-lock');
```
Comment (trasher) [inline src/Glpi/Inventory/Conf.php:394]: I don't think it's relevant to use this method here; we're not creating a tab; plus that may cause issues when one will try to convert this page to a Twig template.

File: tests/web/Glpi/Inventory/Request.php [hunk 25]
```diff
@@ -40,7 +40,7 @@
 use GuzzleHttp\Exception\RequestException;
 use GuzzleHttp\Psr7\Response;
 
-class Request extends \GLPITestCase
+class Request extends \DBTestCase
```
Comment (trasher) [inline tests/web/Glpi/Inventory/Request.php:43]: Is this change really required?

File: src/Glpi/Inventory/Conf.php (hunk 24 above)
Comment (stonebuzz) [reply src/Glpi/Inventory/Conf.php:394]: It's just that we've done this for all the other options

```
echo \Item_Disk::createTabEntry(\Item_Disk::getTypeName(Session::getPluralNumber()), 0, \Item_Disk::getType());
echo \Software::createTabEntry(\Software::getTypeName(Session::getPluralNumber()), 0, \Software::getType());
echo \Monitor::createTabEntry(\Monitor::getTypeName(Session::getPluralNumber()), 0, \Monitor::getType());
 echo \Printer::createTabEntry(\Printer::getTypeName(Session::getPluralNumber()), 0, \Printer::getType());
....
```

Comment (stonebuzz) [reply src/Glpi/Inventory/Conf.php:394]: However, instead of using `CommonGLPI`, we could use `Oauthclient`.

```php
echo \OAuthClient::createTabEntry(\OAuthClient::getTypeName(Session::getPluralNumber()), 0, \OAuthClient::getType());
```

Comment (trasher) [reply src/Glpi/Inventory/Conf.php:394]: Sounds strange, but OK... That part must be migrated to a twig template anyway.

Comment (trasher) [reply src/Glpi/Inventory/Conf.php:394]: > However, instead of using `CommonGLPI`, we could use `Oauthclient`.
> 
> ```
> echo \OAuthClient::createTabEntry(\OAuthClient::getTypeName(Session::getPluralNumber()), 0, \OAuthClient::getType());
> ```

I do not get the point here, bu I think it's not necessary to change. As I said, this must be migrated to Twig, and therefore won't use `createTabEntry()` anymore.

Also, I do not know what `OAuthClient::getTypeName()` gives, but it should certainly be used only in oauth context, which is not the case here.

File: tests/web/Glpi/Inventory/Request.php (hunk 25 above)
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:43]:
```suggestion
class Request extends \GLPITestCase
```

revert to `GLPITestCase`

File: tests/web/Glpi/Inventory/Request.php [hunk 26]
```diff
@@ -207,4 +207,204 @@ public function testPrologRequest()
         );
         $this->checkXmlResponse($res, '<PROLOG_FREQ>24</PROLOG_FREQ><RESPONSE>SEND</RESPONSE>', 200);
     }
+
+    public function testAuthBasic()
+    {
+        global $DB;
+        $basic_auth_password = "a_password";
+        $basic_auth_login = "a_login";
+        $this->login();
```
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:216]:
```suggestion
```

File: tests/web/Glpi/Inventory/Request.php [hunk 27]
```diff
@@ -207,4 +207,204 @@ public function testPrologRequest()
...
+        $this->boolean($conf->saveConf(
+            [
+                "enabled_inventory" => true,
+                'auth_required' => \Glpi\Inventory\Conf::BASIC_AUTH,
+                'basic_auth_login' => $basic_auth_login,
+                'basic_auth_password' => $basic_auth_password
+            ]
+        ))->isTrue();
+        $DB->commit();
+        $this->logout();
```
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:227]:
```suggestion
```

File: tests/web/Glpi/Inventory/Request.php [hunk 28]
```diff
@@ -207,4 +207,204 @@ public function testPrologRequest()
...
+        $this->checkXmlResponse($res, '<PROLOG_FREQ>24</PROLOG_FREQ><RESPONSE>SEND</RESPONSE>', 200);
+    }
+
+    public function testAuthBasicMalformed()
+    {
+        global $DB;
+        $basic_auth_password = "a_password";
+        $basic_auth_login = "a_login";
+
+        $this->login();
```
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:278]:
```suggestion
```

File: tests/web/Glpi/Inventory/Request.php [hunk 29]
```diff
@@ -207,4 +207,204 @@ public function testPrologRequest()
...
+        $this->boolean($conf->saveConf(
+            [
+                "enabled_inventory" => true,
+                'auth_required' => \Glpi\Inventory\Conf::BASIC_AUTH,
+                'basic_auth_login' => $basic_auth_login,
+                'basic_auth_password' => $basic_auth_password
+            ]
+        ))->isTrue();
+        $DB->commit();
+        $this->logout();
```
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:289]:
```suggestion
```

File: tests/web/Glpi/Inventory/Request.php [hunk 30]
```diff
@@ -207,4 +207,204 @@ public function testPrologRequest()
...
+        $this->string((string)$response->getBody())->isEqualTo('{"status":"error","message":"Access denied. Wrong login or password for basic authentication.","expiration":24}');
+    }
+
+    public function testAuthBasicWithFakeCredential()
+    {
+        global $DB;
+        $basic_auth_password = "a_password";
+        $basic_auth_login = "a_login";
+
+        $this->login();
```
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:348]:
```suggestion
```

File: tests/web/Glpi/Inventory/Request.php [hunk 31]
```diff
@@ -207,4 +207,204 @@ public function testPrologRequest()
...
+        $this->boolean($conf->saveConf(
+            [
+                "enabled_inventory" => true,
+                'auth_required' => \Glpi\Inventory\Conf::BASIC_AUTH,
+                'basic_auth_login' => $basic_auth_login,
+                'basic_auth_password' => $basic_auth_password
+            ]
+        ))->isTrue();
+        $DB->commit();
+        $this->logout();
```
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:359]:
```suggestion
```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: tests/web/Glpi/Inventory/Request.php (hunk 25 above)
Comment (stonebuzz) [reply tests/web/Glpi/Inventory/Request.php:43]: Yes because we need to login before updating the config  (`login()` is available from `DBTestCase`)
because `saveConf` check for `Config::canUpdate()`
example : 
```php
 $this->login();
 $this->boolean($conf->saveConf(
            [
                "enabled_inventory" => true,
                'auth_required' => \Glpi\Inventory\Conf::BASIC_AUTH,
                'basic_auth_login' => $basic_auth_login,
                'basic_auth_password' => $basic_auth_password
            ]
        ))->isTrue();
```

File: tests/web/Glpi/Inventory/Request.php [hunk 32]
```diff
@@ -207,4 +207,204 @@ public function testPrologRequest()
...
+        $conf = new \Glpi\Inventory\Conf();
+        $this->boolean($conf->saveConf(
+            [
+                "enabled_inventory" => true,
+                'auth_required' => \Glpi\Inventory\Conf::BASIC_AUTH,
+                'basic_auth_login' => $basic_auth_login,
+                'basic_auth_password' => $basic_auth_password
+            ]
+        ))->isTrue();
+        $DB->commit();
```
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:226]:
```suggestion
```

File: tests/web/Glpi/Inventory/Request.php [hunk 33]
```diff
@@ -207,4 +207,204 @@ public function testPrologRequest()
         );
         $this->checkXmlResponse($res, '<PROLOG_FREQ>24</PROLOG_FREQ><RESPONSE>SEND</RESPONSE>', 200);
     }
+
+    public function testAuthBasic()
+    {
+        global $DB;
```
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:213]:
```suggestion
```

File: tests/web/Glpi/Inventory/Request.php [hunk 34]
```diff
@@ -207,4 +207,204 @@ public function testPrologRequest()
...
+                  '<QUERY>PROLOG</QUERY>' .
+                '</REQUEST>'
+            ]
+        );
+        $this->checkXmlResponse($res, '<PROLOG_FREQ>24</PROLOG_FREQ><RESPONSE>SEND</RESPONSE>', 200);
+    }
+
+    public function testAuthBasicMalformed()
+    {
+        global $DB;
```
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:274]:
```suggestion
```

File: tests/web/Glpi/Inventory/Request.php [hunk 35]
```diff
@@ -207,4 +207,204 @@ public function testPrologRequest()
...
+        $conf = new \Glpi\Inventory\Conf();
+        $this->boolean($conf->saveConf(
+            [
+                "enabled_inventory" => true,
+                'auth_required' => \Glpi\Inventory\Conf::BASIC_AUTH,
+                'basic_auth_login' => $basic_auth_login,
+                'basic_auth_password' => $basic_auth_password
+            ]
+        ))->isTrue();
+        $DB->commit();
```
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:288]:
```suggestion
```

File: tests/web/Glpi/Inventory/Request.php [hunk 36]
```diff
@@ -207,4 +207,204 @@ public function testPrologRequest()
...
+        );
+        $this->object($this->exception)->isInstanceOf(RequestException::class);
+        $this->object($response = $this->exception->getResponse())->isInstanceOf(Response::class);
+        $this->integer($response->getStatusCode())->isEqualTo(401);
+        $this->string((string)$response->getBody())->isEqualTo('{"status":"error","message":"Access denied. Wrong login or password for basic authentication.","expiration":24}');
+    }
+
+    public function testAuthBasicWithFakeCredential()
+    {
+        global $DB;
```
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:344]:
```suggestion
```

File: tests/web/Glpi/Inventory/Request.php [hunk 37]
```diff
@@ -207,4 +207,204 @@ public function testPrologRequest()
...
+        $conf = new \Glpi\Inventory\Conf();
+        $this->boolean($conf->saveConf(
+            [
+                "enabled_inventory" => true,
+                'auth_required' => \Glpi\Inventory\Conf::BASIC_AUTH,
+                'basic_auth_login' => $basic_auth_login,
+                'basic_auth_password' => $basic_auth_password
+            ]
+        ))->isTrue();
+        $DB->commit();
```
Comment (stonebuzz) [inline tests/web/Glpi/Inventory/Request.php:358]:
```suggestion
```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

**Comment by stonebuzz** [comment]
`$DB->commit()` should not be used
I was able to check that the configuration had been saved correctly

File: tests/web/Glpi/Inventory/Request.php (hunk 25 above)
Comment (trasher) [reply tests/web/Glpi/Inventory/Request.php:42]: OK, but DBTestCase is meant to work from the same call, not from an external one - a transaction is created and never commited.
You have to explicitly commit all database change so they are persisted - and you also have to reset properly previous state and commit explicitly too. I'm not sure what else exactly it implies (just check beforeTestsMethod and other in the DbTestCase class.

**Comment by cconard96** [comment]
I think there should be a warning displayed when trying to use basic authentication. It really should never be used, except as a short-term solution for anyone that was handling this at the web server level previously until they are able to switch to using OAuth.

**Comment by trasher** [comment]
> I think there should be a warning displayed when trying to use basic authentication. It really should never be used, except as a short-term solution for anyone that was handling this at the web server level previously until they are able to switch to using OAuth.

If we warn about using an unsecure auth mechanism, we have to warn about using default no auth at all one. I wonder if a better place would be in documentation.

**Review by cconard96 — APPROVED** [review]

**Review by trasher — APPROVED** [review]
